### PR TITLE
[feature/plugin] JWT 

### PR DESCRIPTION
**Pull request by thibaultCha**
This implements JWT validation as discussed in #519. It takes some elements from #401 but with use cases described in #519 only. It supports HMAC SHA-256 signed JWTs only.

Here is what is implemented:

```shell
$ http :8001/apis/{api}/plugins name=jwt
HTTP/1.1 200 OK
...

$ http POST :8001/consumers/{consumer}/jwt
HTTP/1.1 201 Created

{
    "consumer_id": "7bce93e1-0a90-489c-c887-d385545f8f4b",
    "created_at": 1442426001000,
    "id": "bcbfb45d-e391-42bf-c2ed-94e32946753a",
    "key": "a36c3049b36249a3c9f8891cb127243c",
    "secret": "e71829c351aa4242c2719cbfbe671c09"
}
```

Crafting of the JWT is done as follows:

Header:

```json
{
  "typ": "JWT",
  "alg": "HS256"
}
```

Claims:

```json
{
  "iss": "a36c3049b36249a3c9f8891cb127243c", // key from the consumer's jwt secret
  "iat": 1442426454,
  "exp": 1442430054,
  "nbf": 1442426454
}
```

Secret is `e71829c351aa4242c2719cbfbe671c09` as generated by Kong.

Final request with JWT is:

```shell
$ http :8000/{api path} 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhMzZjMzA0OWIzNjI0OWEzYzlmODg5MWNiMTI3MjQzYyIsImV4cCI6MTQ0MjQzMDA1NCwibmJmIjoxNDQyNDI2NDU0LCJpYXQiOjE0NDI0MjY0NTR9.AhumfY35GFLuEEjrOXiaADo7Ae6gt_8VLwX7qffhQN4'
```

Or in querystring:

```shell
$ http ':8000/{api path}?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJhMzZjMzA0OWIzNjI0OWEzYzlmODg5MWNiMTI3MjQzYyIsImV4cCI6MTQ0MjQzMDA1NCwibmJmIjoxNDQyNDI2NDU0LCJpYXQiOjE0NDI0MjY0NTR9.AhumfY35GFLuEEjrOXiaADo7Ae6gt_8VLwX7qffhQN4'
```

If the JWT is verified by Kong, then the request is proxied **untouched** to the upstream service.

##### Customizations

- The querystring parameter is customizable. 
- The users's `key` (which goes in the `iss` claim) is auto-generated if not specified but can be a `POST` parameter if a system needs to migrate to Kong.

##### Improvements

Now, @jasonmotylinski-dowjones, @lucamaraschi, @jerizm, @sinzone, @thefosk as you all took part in this, I would like to have feedback. Does this satisfies your use cases or does it lack a feature? Here is a list of features that this PR does **not** support but could implement:

- Send the **decoded** claims of the JWT to the upstream service (as a JSON payload maybe?).
- Validating other fields of the JWT (`exp`, `nbf`) and reject it if invalid. It currently only validates the signature.
- Implement support for `jti`. Unique JWTs that cannot be replayed.

Since all those are application specific I wanted feedback on whether or not they were required since I think they have value but were not mentioned in #519...

### CODE REVIEW

**Comment by lucamaraschi** [comment]
It looks that the main functionalities are satisfying what I was looking for, but I would add the expiration validation check.

**Comment by jerizm** [comment]
Definitely satisfies our use cases. 

I would also add the exp and nbf checks. Those claims are optional in the JWT spec so if they are included in the JWT they should be validated.

**Comment by tecnobrat** [comment]
I would love to have it decode the claim (or an option of doing so).

Anything with the claim data is custom logic, better solved with the upstream service (authorization vs authentication). If we can have the only piece of our application that needs to know about how to decode the claim be inside kong, it drastically reduces what each upstream service needs to know.  The upstream service can just deal with authorization instead of authentication.

Basically kong would decrypt the claim.  If it validates properly, it passes along the decrypted claim.

File: kong/plugins/jwt/access.lua [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  for _, v in ipairs(conf.uri_param_names) do
+    if uri_parameters[v] then
+      return uri_parameters[v]
+    end
+  end
+
+  local authorization_header = request.get_headers()["authorization"]
+  if authorization_header then
+    local iterator, iter_err = re_gmatch(authorization_header, "\\s*[Bb]earer\\s*(.+)")
```
Comment (tecnobrat) [inline kong/plugins/jwt/access.lua]: Shouldn't the whitespace between Bearer and the token be required?  You don't want to match `BearerToken` or anything weird. `\\s+` instead of `\\s*`

I also question the whitespace at the start of the header, but I'll digress on that one.

I would recommend `"\\s*[Bb]earer\\s+(.+)"

**Comment by jerizm** [comment]
I think decoding a JWT is pretty trivial since it's just a base64 decode so I'd just handle it upstream. 

For our use case the JWT can have more complex properties than just key value pairs for claims so we just deserialize it into a Profile model.

**Comment by tecnobrat** [comment]
@jerizm the JWT is encrypted.  So if you handle that upstream, all of your upstreams need the key.  Unless I'm missing something?

**Comment by thibaultCha** [comment]
The JWT is not encrypted. It is simply signed.

**Comment by lucamaraschi** [comment]
Agree with @thibaultCha 
Is this feature going to land in 0.5.0?

**Comment by tecnobrat** [comment]
Apparently I am crazy and I didn't completely understand JWT.  This PR will work perfectly for our use-case as-is.

**Comment by thibaultCha** [comment]
> Is this feature going to land in 0.5.0?

Yes, which should be release next week. I am currently implementing the exp and nbf check, as well as writing a JWT parser specific to Kong for better maintainability and to avoid decoding twice the JWT as jwtlua forces us to do...

**Comment by thibaultCha** [comment]
The latest commit adds support for `exp` and `nbf` validation. The custom JWT parser should also be more efficient/maintainable and gives us more flexibility.

New config option `claims_to_verify`, empty be default (no checks performed):

```http
$ http :8001/apis/{api}/plugins name=jwt config.claims_to_verify=exp,nbf
HTTP/1.1 201 Created

{
    "api_id": "bc5f6a5a-5c13-4449-c496-6e1ff45192a9",
    "config": {
        "claims_to_verify": [
            "exp",
            "nbf"
        ],
        "uri_param_names": [
            "jwt"
        ]
    },
    "created_at": 1442585600000,
    "enabled": true,
    "id": "c4b34c16-b57f-4921-c9b1-cc5321f1e467",
    "name": "jwt"
}
```

Now both `nbf` and `exp` fields will be required and verified.

**Comment by jerizm** [comment]
Looks great! :+1:

**Comment by vzaramel** [comment]
I don't if this is the right place to ask this question. But would be possible to implement the method described in this post http://nordicapis.com/how-to-control-user-identity-within-microservices/ with this plugin?

**Comment by tecnobrat** [comment]
@vzaramel maybe you could explain what you are looking for out of that 8 page blog post?

This PR is in 0.5.0 which supports validation of the JWT, so if we are looking at this (from the blog post)

>
1) The Reference Token is issued by the Authorization Server. The client sends back when it’s time to call the API.
2) In the middle: The Authorization Server validates the token and responds with a JWT.
3) The JWT is then passed further along in the network.

(1) is handled by you, you need to ensure that happens.
(2) is handled by this PR, it will validate the JWT and ensure the signature is correct
(3) the header is passed upstream untouched, so you can do whatever you want with it (like pull out the user data encoded inside it)

**Comment by vzaramel** [comment]
Thanks for the answer. I'm sorry for not have been especific about the topic.  My question is whether this PR can be the translation service. Sending and receiving the reference token to external user and forwarding a JWT to my internal services.
